README
geximon 0.7

  by Gintautas Miliauskas <gintas@pov.lt>
  Programmers of Vilnius

  Website: http://pov.lt/geximon


>>> Introduction <<<

  geximon is a replacement for eximon, the exim server monitor. The original
  program is functional, but, to cite one of the exim mailing list members,
  "it could use a visit from one of the Learning Channel redecoration shows
  :-)".
  
  geximon is written in Python and is based on Gtk2. The source code is
  available under the GNU General Public Licence. This is a development
  version, so your mileage may vary, however, there should not be any
  serious breakage. If you happen to come across a bug, please take the time
  to drop me an e-mail.
  
  You will need Python[1] (version >= 2.2) and a recent version of pygtk[2].
  Of course, you will need exim[3] (version 3 or 4) installed as well. If
  exim is installed in a non-standard location, you will need to specify the
  paths in geximon preferences.

  [1] http://www.python.org
  [2] http://www.daa.com.au/~james/software/pygtk/
  [3] http://www.exim.org

  This version has been tested with Debian testing and unstable.


>>> Installation <<<

  You may run the program by running `geximon.py` in the geximon
  directory.  Moreover, as this program is packaged using distutils, you can
  easily install it by running `./setup.py install` as the superuser. A
  script named `geximon` will be placed in your path which you can then use
  to invoke the program.

  If you are using a Debian-based operating system, you should use the Debian
  package rather than distutils. The package is available at the website.


  More documentation is available in HTML format in the `doc/` subdirectory.
  It should also be accessible from geximon through the Help menu.


debian/changelog
geximon (0.7.1-1) unstable; urgency=low

  * New upstream release
  * Geximon adopted by David Watson <david@planetwatson.co.uk>
  * Initial Debian upload. Closes: #244210

 -- David Watson <david@planetwatson.co.uk>  Thu, 22 Sep 2005 21:32:47 +0100

geximon (0.7-1) unstable; urgency=low

  * Added a manpage.
  * The GUI should be more responsive while the queue display is being
    updated. Also, large queues should be handled much faster. Thanks to
    Laurent Martelli for input regarding this issue.
  * Now depending on exim4-base rather than exim4. Again thanks to Laurent
    Martelli for pointing this out.
  * Changed architecture from 'any' to 'all'. The package should be
    installable on non-i386 systems now.
  * Miscellaneous small bugfixes, code cleanups and minor feature additions.

 -- Gintautas Miliauskas <gintas@pov.lt>  Sat, 10 Apr 2004 01:06:07 +0300

geximon (0.6-1) unstable; urgency=low

  * Initial Release.

 -- Gintautas Miliauskas <gintas@pov.lt>  Wed, 27 Mar 2004 17:24:58 +0200



test/test.py
#!/usr/bin/env python
"""Geximon unit testing module."""

import unittest
import doctest

import sys
sys.path.insert(0, '..')

class TestWindows(unittest.TestCase):

    def test_createGEximonWindow(self):
        from geximon.geximon import GEximonWindow
        win = GEximonWindow()

    def test_createProcessWindow(self):
        from geximon.geximon import ProcessWindow
        from geximon.preferences import Preferences
        prefs = Preferences()
        w = ProcessWindow(prefs)

    def test_createPopupWindow(self):
        from geximon.widgets import PopupWindow
        w = PopupWindow("title", "text")


class TestPreferences(unittest.TestCase):

    def test_createDialog(self):
        from geximon.preferences import Preferences, PreferencesDialog
        prefs = Preferences()
        dlg = PreferencesDialog(None, prefs)


if __name__ == '__main__':
    suite = unittest.TestSuite()

    from geximon import exim
    suite.addTest(doctest.DocTestSuite(exim))

    suite.addTest(unittest.makeSuite(TestWindows))
    suite.addTest(unittest.makeSuite(TestPreferences))
    unittest.TextTestRunner(verbosity=2).run(suite)


geximon/exim.py
"""Interfacing with exim processes."""

import os
import threading

__metaclass__ = type


def get_output(path='', filename='', args='', use_sudo=False):
    """Run a command and return its output."""
    cmd = (use_sudo and 'sudo ' or '') + os.path.join(path, filename)
    if args:
        cmd += ' ' + args
    stdin, stdouterr = os.popen4(cmd)
    stdin.close()
    try:
        return stdouterr.read().strip()
    finally:
        stdouterr.close()


class LogWatcher:
    """Watch exim logs."""

    def __init__(self, log_dir, mainlog_name, bin_dir, sudo, line_limit=500):
        self.line_limit = line_limit
        self.for_processing = []
        self._valid = True
        self.bin_dir = bin_dir
        self.use_sudo = sudo
        self.open(log_dir, mainlog_name)

    def open(self, log_dir, mainlog_name):
        self.log_dir = log_dir
        self.mainlog_name = mainlog_name
        mainlog_path = os.path.join(log_dir, mainlog_name)
        try:
            self.mainlog = open(mainlog_path)
            self.mainlog_inode = os.fstat(self.mainlog.fileno()).st_ino
            self._valid = True
        except IOError:
            if self._valid:
                # XXX should show a popup
                self.unseen = \
                    [_("Error: could not open the exim log file at `%s`!")
                                                              % mainlog_path]
                self._valid = False
            self.mainlog = None
        else:
            self.unseen = self.mainlog.readlines()[-self.line_limit:]
            self.unseen = [s[:-1] for s in self.unseen]

    def update(self):
        """Read the log file for new entries."""
        mainlog_path = os.path.join(self.log_dir, self.mainlog_name)
        if self.mainlog:
            new_entries = self.mainlog.readlines()
            new_entries = [s[:-1] for s in new_entries]
        else:
            new_entries = []
            self.open(self.log_dir, self.mainlog_name)
            return

        try:
            new_ino = os.stat(mainlog_path).st_ino
        except OSError:
            pass
        else:
            if self.mainlog_inode != new_ino:
                self.open(self.log_dir, self.mainlog_name)

        self.unseen += new_entries
        self.for_processing += new_entries

    def get_unseen(self):
        """Get unseen entries."""
        unseen = self.unseen
        self.unseen = []
        return unseen

    def get_for_processing(self):
        """Get list of lines that have not been processed yet."""
        for_processing = self.for_processing
        self.for_processing = []
        return for_processing

    def runExigrep(self, pattern, literal, all_logs):
        """Run exigrep with the provided pattern.

        `pattern` is a Perl regular expression, or a plain string - if
        `literal` is True. `all_logs` indicates if all logfiles (rather
        than just the latest one) should be scanned.
        """
        filename = os.path.join(self.log_dir, all_logs and '*'
                                              or self.mainlog_name)
        pattern = "'" + pattern + "'"
        return get_output(self.bin_dir, 'exigrep',
                (literal and '-l' or '') + pattern + ' ' + filename)

    def runEximstats(self, args, all_logs):
        """Run eximstats.

        `args` is a string containing the arguments to pass to eximstats,
        all_logs indicates if all logfiles should be scanned.
        """
        filename = os.path.join(self.log_dir, all_logs and '*'
                                              or self.mainlog_name)
        return get_output(self.bin_dir, 'eximstats', args + ' ' + filename)

    def getRejectlog(self):
        """Get the contents of the rejectlog."""
        # XXX this heuristic is unreliable and should be refactored
        filename = os.path.join(self.log_dir, 'rejectlog')
        if not os.path.exists(filename):
            filename = os.path.join(self.log_dir, 'reject.log')
        try:
            f = open(filename)
            text = f.read()
        except IOError:
            text = '' # XXX should indicate an error
        return text


class BackgroundJob:
    """On-demand background job manager.

    BackgroundJob starts a new background thread that goes to sleep.  When it
    is woken up (by calling schedule_update), it performs some background
    processing (do_update) and goes to sleep again.  Any calls to
    schedule_update while the background processing is active are ignored.

    The background thread can be stopped by calling stop.

    Subclasses must override do_update.
    """

    def __init__(self):
        self._quit = False
        self._synch = threading.Event()
        self._thread = threading.Thread(target=self.bg_processing)
        self._thread.setDaemon(True)
        self._thread.start()

    def bg_processing(self):
        """Perform background processing on demand."""
        while not self._quit:
            self._synch.wait()
            if self._quit:
                break
            self.do_update()
            self._synch.clear()

    def schedule_update(self):
        """Wake up the background thread."""
        self._synch.set()

    def stop(self, timeout=1.0):
        """Stop background processing."""
        self._quit = True
        self._synch.set()
        self._thread.join(timeout)

    def do_update(self):
        """Override this method to perform background jobs.

        It is called in a background thread.
        """
        raise NotImplementedError("You need to override do_update in %s"
                                  % self.__class__.__name__)


class QueueManager(BackgroundJob):
    """Exim queue manager.

    Queue manager calls exim and parses its output in a background thread to
    get the status of the exim queue.  Once the status is available, the queue
    manager calls the registered callback.
    """

    def __init__(self, callback, bin_dir, exim_binary, use_sudo):
        """Initialize and start the background thread.

        callback is a callable that takes one argument -- a mapping from
        message IDs to message objects.  It will be called from the background
        thread.

        bin_dir is the path to exim.
        use_sudo indicates if sudo is to be used.
        """
        BackgroundJob.__init__(self)
        self.callback = callback
        self.bin_dir = bin_dir
        self.exim_binary = exim_binary
        self.use_sudo = use_sudo
        self.queue_length = 0

    def do_update(self):
        """Get the current exim queue in a separate thread."""
        data = self.getEximOutput(['-bpr'])

        if (data.find('No such file or directory') != -1
                or data.find(' not found') != -1
                or data.find('exim: permission denied') != -1):
            # complain in case of problems
            m = Message()
            m.frozen, m.time, m.size, m.sender = True, "", "", "",
            # XXX should show a popup
            m.id, m.recipients = 'error', \
                [_("Error invoking `exim -bpr`:\n") + data]
            self.messages = {'error': m}
            self.callback(self.messages)
            return

        data = data.splitlines()
        messages = {}
        recipient_list = False
        message = None
        for line in data:
            if not recipient_list:
                try:
                    message = Message(status_line=line)
                except ValueError:
                    pass # probably an odd status line, ignore it
                else:
                    messages[message.id] = message
                recipient_list = True
            elif line:
                message.add_recipient(line)
            else:
                recipient_list = False
        self.queue_length = len(messages)
        self.callback(messages)

    def getEximOutput(self, args):
        """Invoke exim with arguments, return command output.

        args is a list of strings (the parameters to be passed).
        """
        return get_output(self.bin_dir, self.exim_binary,
                " ".join(args), self.use_sudo)

    def checkOutput(self, output, expected, msg_count, action_name):
        """Check if a string has `expected` as a substring on each line.

        Returns a tuple: (no_errors, report_string).
        """
        lines = output.splitlines()
        errors = filter(lambda s: s.find(expected) == -1, lines)
        if not errors:
            words = msg_count > 1 and 'messages have' or 'message has'
            success_report = _("%d %s been %s.") \
                    % (msg_count, words, action_name)
            return (True, success_report)
        else:
            return (False, "\n".join(errors))

    # -- general actions --

    def runQueue(self):
        """Run the queue now."""
        # XXX a little dirty, but will do for now
        cmd = os.path.join(self.bin_dir, self.exim_binary) + ' -q &'
        if self.use_sudo:
            cmd = 'sudo ' + cmd
        os.system(cmd)
        return _("Spawning a queue runner in the background.")

    def getConfiguration(self):
        """Get all exim configuration options."""
        return self.getEximOutput(['-bP'])

    # -- message actions --
    # methods must return a tuple: (action_successful, message)

    def getMessageBody(self, id):
        """Get the body of the message with the given ID."""
        return (True, self.getEximOutput(['-Mvb', id]))

    def getMessageHeaders(self, id):
        """Get the headers of the message with the given ID."""
        return (True, self.getEximOutput(['-Mvh', id]))

    def getMessageLog(self, id):
        """Get the log of the message with the given ID."""
        return (True, self.getEximOutput(['-Mvl', id]))

    def removeMessages(self, ids):
        """Remove messages with given IDs."""
        output = self.getEximOutput(['-Mrm'] + ids)
        return self.checkOutput(output,
                "has been removed", len(ids), _("removed"))

    def freezeMessages(self, ids):
        """Freeze messages with given IDs."""
        output = self.getEximOutput(['-Mf'] + ids)
        return self.checkOutput(output,
                "is now frozen", len(ids), _("frozen"))

    def thawMessages(self, ids):
        """Unfreeze messages with given IDs."""
        output = self.getEximOutput(['-Mt'] + ids)
        return self.checkOutput(output,
                "is no longer frozen", len(ids), _("thawed"))

    def deliverMessages(self, ids):
        """Force the delivery of messages with given IDs."""
        return (True, self.getEximOutput(['-M'] + ids))

    def giveUpMessages(self, ids):
        """Give up trying to deliver messages with given IDs."""
        return (True, self.getEximOutput(['-Mg'] + ids))

    def addRecipients(self, id, recipients):
        """Add more recipients to a message with a given ID."""
        output = self.getEximOutput(['-Mar', id] + recipients.split())
        return self.checkOutput(output,
                "has been modified", 1, _("modified"))

    def editSender(self, id, sender):
        """Change the sender of a message with a given ID."""
        output = self.getEximOutput(['-Mes', id, sender])
        return self.checkOutput(output,
                "has been modified", 1, _("modified"))

    def markAllDelivered(self, ids):
        """Mark all recipients of messages with given IDs as delivered."""
        output = self.getEximOutput(['-Mmad'] + ids)
        return self.checkOutput(output,
                "has been modified", len(ids), _("marked as delivered"))


class Message:
    """A queued message.

    >>> line = '22h  1.2K 1AttSk-0002qB-00 <> *** frozen ***'
    >>> m = Message(line)
    >>> m.id, m.frozen, m.time, m.sender, m.size, m.recipients
    ('1AttSk-0002qB-00', True, '22h', '<>', '1.2K', [])
    >>> m.add_recipient(' foobar\\n')
    >>> m.recipients
    ['foobar']
    """

    def __init__(self, status_line=None):
        """Extract information from a message status line returned by exim."""
        if not status_line is None:
            parts = status_line.split()
            if len(parts) < 4:
                raise ValueError(_("Invalid status line: ") + status_line)
            self.id = parts[2]
            self.frozen = "frozen" in parts
            self.time = parts[0]
            self.size = parts[1]
            self.sender = parts[3]
            self.recipients = []

    def add_recipient(self, line):
        """Add a recipient to the message."""
        self.recipients.append(line.strip())

    def __eq__(self, other):
        return (self.id == other.id and self.frozen == other.frozen and
                self.time == other.time and self.size == other.size and
                self.sender == other.sender and
                self.recipients == other.recipients)

    def __ne__(self, other):
        return not self.__eq__(other)


class ProcessManager(BackgroundJob):
    """Exim process manager.

    Process manager calls exiwhat(8) and parses its output in a background
    thread to get the status of exim processes.  Once the status is available,
    the process manager calls the registered callback.
    """

    def __init__(self, callback, bin_dir, use_sudo):
        """Start the background thread.

        callback is a callable that takes one argument -- a mapping from
        process IDs to status messages.  It will be called from the background
        thread.

        bin_dir is the path to exim binaries.
        use_sudo indicates if sudo is to be used.
        """
        BackgroundJob.__init__(self)
        self.callback = callback
        self.bin_dir = bin_dir
        self.use_sudo = use_sudo

    def do_update(self):
        """Collect data from exiwhat in a separate thread."""
        data = get_output(self.bin_dir, 'exiwhat', use_sudo=self.use_sudo)

        processes = {}
        if (data.find('Permission denied') != -1 or
            data.find('Operation not permitted') != -1):
            # XXX need to be more verbose
            status = _("Permission problems!")
        elif data.find('No exim process data') != -1:
            status = _("No exim processes are currently running.")
        elif data == '':
            status = _("No output from exiwhat!")
        else:
            status = ""
            for line in data.splitlines():
                try:
                    id, info = line.split(None, 1)
                    id = int(id)
                    processes[id] = info.strip()
                except ValueError:
                    # probably trying to parse an error message
                    status = _("Error processing exiwhat output line: ") + line
            if not status:
                process_word = (len(processes) > 1 and _("processes")
                                                    or _("process"))
                status = _("%d exim %s.") % (len(processes), process_word)

        self.callback(processes, status)


geximon/geximon.py
#!/usr/bin/env python
"""
    Copyright (C) 2004  Gintautas Miliauskas <gintas@pov.lt>,
                        Programmers of Vilnius

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
"""

VERSION = "0.7.1"

import gettext
gettext.install('geximon')

import gtk
import gobject

from exim import LogWatcher, QueueManager
from gtkhelpers import AlertDialog
from gtkhelpers import framed, scrolled, show_help
from widgets import LogWidget, ProcessWidget, QueueWidget
from widgets import PopupWindow, ExigrepDialog, EximstatsDialog
from preferences import Preferences, PreferencesDialog
from plotter import Plotter


class GEximonWindow(gtk.Window):
    """The main geximon window."""

    def __init__(self):
        gtk.Window.__init__(self, gtk.WINDOW_TOPLEVEL)
        self.set_title('gEximon')
        self.connect('delete-event', self.quit)

        self.prefs = prefs = Preferences()
        prefs.load()
        prefs.subscribe(self.apply_prefs)

        self.logwatcher = LogWatcher(prefs.log_dir, prefs.mainlog_name,
                prefs.bin_dir, prefs.use_sudo)

        # callback is None, it will be specified by QueueWidget
        self.queue_mgr = QueueManager(None,
                prefs.bin_dir, prefs.exim_binary, prefs.use_sudo)

        self._setUpIcons()

        # create menu bar
        self._setUpMenu()

        # create status bar
        self.statusbar = gtk.Statusbar()
        self.statusbar.push(0, '')

        # set up widgets
        self.plotter = Plotter(self.logwatcher, self.queue_mgr, prefs)
        self.log_widget = LogWidget(self.logwatcher, prefs)
        self.queue_widget = QueueWidget(self,
                self.logwatcher, self.queue_mgr, prefs)

        self.process_window = ProcessWindow(prefs)
        def processWindowClosed(*ignored_arguments):
            self.prefs.show_process_list = False
            self.prefs.notify()
            return True # prevent the window from getting destroyed
        self.process_window.connect('delete-event', processWindowClosed)

        self._layOutWidgets()
        prefs.notify()

    def _setUpIcons(self):
        try:
            icon16x16 = gtk.gdk.pixbuf_new_from_file(
                    '/usr/share/pixmaps/geximon-16x16.xpm')
            icon32x32 = gtk.gdk.pixbuf_new_from_file(
                    '/usr/share/pixmaps/geximon-32x32.xpm')
        except gobject.GError:
            pass
        else:
            gtk.window_set_default_icon_list(icon32x32, icon16x16)

    def _setUpMenu(self):
        """Set up the main menubar."""

        def runQueue(*ignored_arguments):
            self.queue_mgr.runQueue()
            if self.prefs.report_success:
                AlertDialog(self, _("A queue runner has been started."),
                    _("A new exim instance has been started in the background."
                      " It will process the queue and then terminate."))

        def runExigrep(*ignored_arguments):
            dialog = ExigrepDialog(self)
            dialog.go(self.logwatcher)

        def runEximstats(*ignored_arguments):
            dialog = EximstatsDialog(self)
            dialog.go(self.logwatcher)

        def getRejectlog(*ignored_arguments):
            text = self.logwatcher.getRejectlog()
            PopupWindow(_("Reject log"), text).show_all()

        def getConfig(*ignored_arguments):
            text = self.queue_mgr.getConfiguration()
            PopupWindow(_("Exim configuration"), text).show_all()

        def scrollLogToggled(action, menuitem):
            """Automatically scroll the log on new data."""
            self.prefs.track_log = menuitem.get_active()
            self.prefs.notify()

        def processDisplayToggled(action, menuitem):
            """Show or hide the process list window."""
            self.prefs.show_process_list = menuitem.get_active()
            self.prefs.notify()

        def plotDisplayToggled(action, menuitem):
            """Show or hide the plot area."""
            self.prefs.show_plotter = menuitem.get_active()
            self.prefs.notify()

        def statusbarDisplayToggled(action, menuitem):
            """Show or hide the statusbar."""
            self.prefs.show_statusbar = menuitem.get_active()
            self.prefs.notify()

        def helpContents(action, menuitem):
            """Show the help contents."""
            show_help()

        accel_group = gtk.AccelGroup()
        item_factory = gtk.ItemFactory(gtk.MenuBar, "<main>", accel_group)
        item_factory.create_items([
                (_("/_Geximon"), None, None, 1, '<Branch>'),
                (_("/Geximon/<tearoff>"), None, None, 0, '<Tearoff>'),
                (_("/Geximon/_Preferences"), '<control>P',
                    self.editPreferences,
                    10, '<StockItem>', gtk.STOCK_PREFERENCES),
                (_("/Geximon/<sep1>"), None, None, 0, '<Separator>'),
                (_("/Geximon/_Quit"), '<control>Q', self.quit,
                    11, '<StockItem>', gtk.STOCK_QUIT),
                (_("/_Exim"), None, None, 2, '<Branch>'),
                (_("/Exim/<tearoff>"), None, None, 0, '<Tearoff>'),
                (_("/Exim/Spawn _Queue Runner"), '<control>R', runQueue, 20,
                    None),
                (_("/Exim/<sep1>"), None, None, 0, '<Separator>'),
                (_("/Exim/Exi_grep"), None, runExigrep, 21, None),
                (_("/Exim/Exim_stats"), None, runEximstats, 22, None),
                (_("/Exim/<sep2>"), None, None, 0, '<Separator>'),
                (_("/Exim/See _Rejectlog"), None, getRejectlog, 23, None),
                (_("/Exim/<sep3>"), None, None, 0, '<Separator>'),
                (_("/Exim/Get _Configuration"), None, getConfig, 23, None),
                (_("/_View"), None, None, 30, '<Branch>'),
                (_("/View/<tearoff>"), None, None, 0, '<Tearoff>'),
                (_("/View/_Plots"), '<control>G', plotDisplayToggled,
                    31, "<CheckItem>"),
                (_("/View/_Statusbar"), None, statusbarDisplayToggled,
                    32, "<CheckItem>"),
                (_("/View/<sep1>"), None, None, 0, '<Separator>'),
                (_("/View/_Process List"), '<control>L', processDisplayToggled,
                    33, "<CheckItem>"),
                (_("/View/<sep2>"), None, None, 0, '<Separator>'),
                (_("/View/_Scroll log on new data"), '<control>S',
                    scrollLogToggled, 34, "<CheckItem>"),
                (_("/_Help"), None, None, 90, '<LastBranch>'),
                (_("/Help/<tearoff>"), None, None, 0, '<Tearoff>'),
                (_("/Help/_Contents"), None, helpContents, 91, None),
                (_("/Help/_About"), None, self.about, 92, None),
                ])
        self.add_accel_group(accel_group)
        self.menubar = item_factory.get_widget('<main>')

        self.show_plotter_menu_item = item_factory.get_item_by_action(31)
        self.show_statusbar_menu_item = item_factory.get_item_by_action(32)
        self.process_list_menu_item = item_factory.get_item_by_action(33)
        self.track_log_menu_item = item_factory.get_item_by_action(34)

        # need to store the reference so it doesn't get garbage collected
        self._item_factory = item_factory

    def _layOutWidgets(self):
        self.pane = gtk.VPaned()
        self.pane.pack1(framed(scrolled(self.log_widget)), resize=True)
        self.pane.pack2(framed(scrolled(self.queue_widget)), resize=True)

        self.vbox = gtk.VBox()
        self.vbox.set_border_width(2)
        self.vbox.pack_start(framed(self.plotter), expand=False)
        self.vbox.pack_end(self.pane)

        self.vbox2 = gtk.VBox()
        self.vbox2.pack_start(self.menubar, expand=False)
        self.vbox2.pack_start(self.vbox, expand=True)
        self.vbox2.pack_end(self.statusbar, expand=False)
        self.add(self.vbox2)
        if self.prefs.remember_sizes:
            self.set_default_size(
                    self.prefs.window_width, self.prefs.window_height)
            self.pane.set_position(self.prefs.divider_position)
        else:
            self.set_default_size(550, 600)

    def quit(self, *ignored_arguments):
        self.save_preferences()
        # it would be nice if the following were called automatically
        self.process_window.process_widget.cleanup()
        gtk.mainquit()
        return False

    def save_preferences(self):
        self.prefs.window_width, self.prefs.window_height = self.get_size()
        self.prefs.divider_position = self.pane.get_position()
        self.prefs.wrap_log = self.log_widget.get_wrap_mode()
        self.prefs.save()

    def about(self, *ignored_arguments):
        """Show the about dialog."""
        about_dialog = AboutDialog(self)
        about_dialog.go()

    def editPreferences(self, *ignored_arguments):
        """Show the preferences dialog."""
        pref_dialog = PreferencesDialog(self, self.prefs)
        pref_dialog.show_all()
        response = pref_dialog.run()
        if response == gtk.RESPONSE_OK:
            pref_dialog.apply(self.prefs)
        pref_dialog.hide()

    def apply_prefs(self, prefs):
        self.process_list_menu_item.set_active(prefs.show_process_list)
        self.show_plotter_menu_item.set_active(prefs.show_plotter)
        self.show_statusbar_menu_item.set_active(prefs.show_statusbar)
        if prefs.show_statusbar:
            self.statusbar.show_all()
        else:
            self.statusbar.hide()
        self.track_log_menu_item.set_active(prefs.track_log)


class ProcessWindow(gtk.Window):
    """A window that contains the exim process list."""

    def __init__(self, prefs):
        gtk.Window.__init__(self, gtk.WINDOW_TOPLEVEL)
        self.set_default_size(500, 200)
        self.set_title(_("Exim processes"))

        self.statusbar = gtk.Statusbar()
        self.statusbar.push(0, '')
        self.process_widget = ProcessWidget(self.statusbar, prefs)

        vbox = gtk.VBox()
        vbox.pack_start(scrolled(self.process_widget))
        vbox.pack_end(self.statusbar, expand=False)

        self.add(framed(vbox))

        prefs.subscribe(self.apply_prefs)

    def apply_prefs(self, prefs):
        if prefs.show_process_list:
            self.process_widget.update()
            self.show_all()
        else:
            self.hide()


class AboutDialog(gtk.Dialog):
    """The About dialog."""

    def __init__(self, main_window):
        gtk.Dialog.__init__(self, _("About gEximon"), main_window,
                flags=(gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT),
                buttons=(gtk.STOCK_OK, gtk.RESPONSE_OK))
        self.set_resizable(False)
        self.set_border_width(12)
        self.vbox.set_spacing(18)

        text = '<span weight="bold" size="larger">gEximon v%s</span>' % VERSION
        title = gtk.Label(text)
        title.set_property('use-markup', True)
        self.vbox.add(title)

        text = _("A GNOME monitor for the exim mail transport agent.")
        about = gtk.Label(text)
        self.vbox.add(about)

        text = ("<i>" + _("Author") + ": " +
                "Gintautas Miliauskas &lt;gintas@pov.lt&gt;\n" +
                "Programmers of Vilnius, 2004\n" +
		"Now maintained by David Watson &lt;david@planetwatson.co.uk&gt;</i>")
        author = gtk.Label(text)
        author.set_property('use-markup', True)
        self.vbox.add(author)

    def go(self):
        self.show_all()
        self.run()
        self.destroy()


def main():
    mainwin = GEximonWindow()
    mainwin.show_all()

    # XXX these should be moved somewhere else
    if not mainwin.prefs.show_plotter:
        mainwin.plotter.set_visible(False)
    if not mainwin.prefs.show_statusbar:
        mainwin.statusbar.hide()
    gtk.gdk.threads_init()
    gtk.main()


if __name__ == '__main__':
    main()
